interface/ceed-operator.c

```diff
@@ -298,6 +298,187 @@ int CeedOperatorAssembleLinearQFunction(CeedOperator op, CeedVector *assembled,
+  CeedTransposeMode lmodeout;
+  for (CeedInt i=0; i<qf->numoutputfields; i++)
+    if (op->outputfields[i]->vec == CEED_VECTOR_ACTIVE) {
+      ierr = CeedOperatorFieldGetBasis(op->outputfields[i], &basisout);
+      CeedChk(ierr);
+      ierr = CeedOperatorFieldGetElemRestriction(op->outputfields[i], &rstrout);
+      CeedChk(ierr);
+      ierr = CeedOperatorFieldGetLMode(op->outputfields[i], &lmodeout);
+      CeedChk(ierr);
+      CeedEvalMode emode;
+      ierr = CeedQFunctionFieldGetEvalMode(qf->outputfields[i], &emode);
+      CeedChk(ierr);
+      switch (emode) {
+      case CEED_EVAL_NONE:
+      case CEED_EVAL_INTERP:
+        ierr = CeedRealloc(numemodeout + 1, &emodeout); CeedChk(ierr);
+        emodeout[numemodeout] = emode;
+        numemodeout += 1;
+        break;
+      case CEED_EVAL_GRAD:
+        ierr = CeedRealloc(numemodeout + dim, &emodeout); CeedChk(ierr);
+        for (CeedInt d=0; d<dim; d++)
+          emodeout[numemodeout+d] = emode;
+        numemodeout += dim;
+        break;
+      case CEED_EVAL_WEIGHT:
+      case CEED_EVAL_DIV:
+      case CEED_EVAL_CURL:
+        break; // Caught by QF Assembly
+      }
+    }
+
+  // Create diagonal vector
+  CeedVector elemdiag;
+  ierr = CeedElemRestrictionCreateVector(rstrin, assembled, &elemdiag);
+  CeedChk(ierr);
+
+  // Assemble element operator diagonals
+  CeedScalar *elemdiagarray, *assembledqfarray;
+  ierr = CeedVectorSetValue(elemdiag, 0.0); CeedChk(ierr);
+  ierr = CeedVectorGetArray(elemdiag, CEED_MEM_HOST, &elemdiagarray);
+  CeedChk(ierr);
+  ierr = CeedVectorGetArray(assembledqf, CEED_MEM_HOST, &assembledqfarray);
+  CeedChk(ierr);
+  CeedInt nelem, nnodes, nqpts;
+  ierr = CeedElemRestrictionGetNumElements(rstrin, &nelem); CeedChk(ierr);
+  ierr = CeedBasisGetNumNodes(basisin, &nnodes); CeedChk(ierr);
+  ierr = CeedBasisGetNumQuadraturePoints(basisin, &nqpts); CeedChk(ierr);
+  // Compute the diagonal of B^T D B
+  // Each node, qpt pair
+  for (CeedInt n=0; n<nnodes; n++)
+    for (CeedInt q=0; q<nqpts; q++) {
+      CeedInt dout = -1;
+      // Each basis eval mode pair
+      for (CeedInt eout=0; eout<numemodeout; eout++) {
+        CeedScalar bt = 1.0;
+        if (emodeout[eout] == CEED_EVAL_GRAD)
+          dout += 1;
+        ierr = CeedBasisGetValue(basisout, &bt, emodeout[eout], dout, n, q);
+        CeedChk(ierr);
+        CeedInt din = -1;
+        for (CeedInt ein=0; ein<numemodein; ein++) {
+          CeedScalar b = 0.0;
+          if (emodein[ein] == CEED_EVAL_GRAD)
+            din += 1;
+          ierr = CeedBasisGetValue(basisin, &b, emodein[ein], din, n, q);
+          CeedChk(ierr);
+          // Each element and component
+          for (CeedInt e=0; e<nelem; e++)
+            for (CeedInt cout=0; cout<ncomp; cout++) {
+            CeedScalar db = 0.0;
+              for (CeedInt cin=0; cin<ncomp; cin++)
+                db +=
+                   assembledqfarray[e*numemodein*ncomp*numemodeout*ncomp*nqpts +
```

Comment:
Note: This is largely just a matter of getting values from the bases and the assembled diagonal and multiplying them together. I am computing the basis values once and reusing them across all elements, though that does mean I am jumping around in the QFunction data quite a bit. I can get more regular access to the QFunction data, at the cost of having to recompute the basis values.